platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,310 @@
+Another benefit of this approach is that adoption trigger becomes very simple (app sets the roster) with delineated
+responsibilities between the app and the platform.
+
+Some edge cases worth considering include:
+
+1. multiple concurrent roster submission - has a potential for introducing race conditions. We will prevent this by
+   guaranteeing immutability on rosters.
+2. Size control on the map of rosters - we certainly don’t want this map to grow infinitely so insertion of rosters will
+   be controlled, with clear rules for removal of unused rosters - the acceptance of a new Candidate Roster
+   will invalidate and remove the current candidate roster.
+
+###### Roster API - Implementation
+
+Implementation of some new TSS components has been proposed.
+See [link](https://www.notion.so/TSS-Platform-Architecture-04b15df371ba4b1d848360542e05a030?pvs=21).
+
+It is **important** to note that before the TSS protocol is implemented and launched, the Platform will adopt a
+Candidate Roster during any software upgrade if the Candidate Roster is present in the state. Only after the TSS
+protocol is launched will we be able to add an extra condition to verify that the Candidate Roster is ready to be
+adopted.
+
+### Core Behaviors, in summary
+
+- Roster Creation: App will create a Candidate Roster from the Candidate Address Book (CAB) and set it in the state.
+- Roster Submission: App will trigger roster adoption by setting the `candidateRosterHash` field in the PlatformState.
+- Roster Adoption: The platform will adopt the last submitted Candidate Roster when it is ready, based on specifications
+  outlined in the TSS Ledger ID Proposal (referenced under related Proposals).
+- Roster Replacement: If a new CAB is submitted before the previous one is adopted, the corresponding new Candidate
+  Roster will replace the old one.
+
+### Public API
+
+A new method will be added to the platform API to allow the App submit a Candidate Roster to the platform:
+
+```java
+//in State
+
+void setCandidateRoster(@NonNull final Roster candidateRoster);
+```
+
+### Configuration
+
+The `config.txt` file will no longer be used for storing the address book in existing networks. It will only be used for
+the genesis of new networks.
+
+### Data storage
+
+All the data related to Active and Candidate Roster will be stored in the State.
+There will not be any other separate artifacts stored elsewhere (e.g., directly on disk.).
+The only artifact that will continue to be stored on disk separately from the State is the `config.txt`, which, as
+explained in *Bootstrapping Genesis for a brand-new network* section below, will only ever be used once during a genesis
+of a brand-new network. The `config.txt` file, or at least its part that describes the address book components, will
+never be used in the life cycle of an existing network after that.
+
+### Startup procedure (pseudo-code)
+
+The pseudocode for the startup procedure will look as follows:
+
+```java
```

Comment:
IMO, this block of pseudocode isn't adding anything to the proposal. In reality, the stuff being done in this code block will happen inside various components, and on various threads. 

I'm generally in favor of expressing things with code, but in this case, I think attempting to render the process as a linear sequence of events gives the wrong impression, and won't really be useful when it comes time to actually implement this.

Reply:
@anthony-swirldslabs, thoughts?

Reply:
This pseudocode was written to answer a few specific questions explicitly asked in the original problem statement at https://github.com/hashgraph/hedera-services/issues/13775 (see the tasks in the description of the issue.)

Also, I believe that the code as written would be running during the startup, on a single thread, in a single place, specifically around here: https://github.com/hashgraph/hedera-services/blob/d1f97ac28ec829cfb0e5d1f2ac5e10675a95af7c/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/builder/PlatformBuilder.java#L505-L560

It does indeed call a few methods in other components, but 99% of this pseudocode itself belongs to this single startup method implementation directly.

However, if the problem statement has changed and those questions no longer require any answers in the design proposal, then the code can be removed.

Reply:
The definition of `Public API` is not just what impacts the broader hedera ecosystem, but also anything that impacts Services.   In that sense, the `PlatformBuilder` and whole bootstrap process of instantiating a platform is public API when we do `inversion of control`.   Services needs to know where to set a `genesisRoster`.    Whatever process that is, it is `Public API` and needs to be documented clear enough for people in Services to call it properly.   That may mean that pseudocode here isn't enough, that we need to show the modified interfaces.

Reply:
is the interface definition under the `Roster Public API` sufficient? What else do you think would be helpful to specify here without going into concrete implementation detail?

Reply:
If I was a services developer with only this document in hand and I was responsible for changing the services code to give the platform a genesis roster at startup, would I be able to do it correctly or figure it out?  How much time would it take me?